### PR TITLE
PLAT-1376 Improved WorkflowItemMessageRowsLoader regarding Auto Transitions

### PR DESCRIPTION
**Pull request by oliver-richers**
https://github.com/softicar/platform/assets/83839745/7939fe02-db41-4e7f-b4c0-92ccc1ae1124

### CODE REVIEW

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java [hunk 1]
```diff
@@ -26,16 +29,41 @@ public WorkflowItemMessageRowsLoader(AGWorkflowItem item) {
...
+		rows.addAll(loadAutoTransitionSourceRows());
+		rows.addAll(loadItemLogs());
+
 		Collections.sort(rows);
+		removedDuplicates(rows);
 		fillNodeAndIndex(rows);
 		return rows;
 	}
 
+	private void removedDuplicates(List<WorkflowItemMessageRow> rows) {
```
Comment (alexander-schmidt) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java:46]:
```suggestion
	private void removeDuplicates(List<WorkflowItemMessageRow> rows) {
```

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java [hunk 2]
```diff
@@ -26,16 +29,41 @@ public WorkflowItemMessageRowsLoader(AGWorkflowItem item) {
...
-		rows.addAll(loadItemLogs());
-		rows.addAll(loadTransitionRows());
 		rows.addAll(loadMessageRows());
+		rows.addAll(loadTaskExecutionRows());
 		rows.addAll(loadDelegationRows());
+		rows.addAll(loadAutoTransitionSourceRows());
+		rows.addAll(loadItemLogs());
+
 		Collections.sort(rows);
+		removedDuplicates(rows);
```
Comment (alexander-schmidt) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java:41]:
```suggestion
		removeDuplicates(rows);
```

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java [hunk 3]
```diff
@@ -26,16 +29,41 @@ public WorkflowItemMessageRowsLoader(AGWorkflowItem item) {
 
 	public List<WorkflowItemMessageRow> loadAllRows() {
 
+		// please note, loading order is important
```
Comment (alexander-schmidt) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java:32]:
```suggestion
		// Loading order is important!
```

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRow.java [hunk 4]
```diff
@@ -94,7 +94,6 @@ public int compareTo(WorkflowItemMessageRow other) {
 
 		return Comparator//
 			.comparing(WorkflowItemMessageRow::getTransaction)
-			.thenComparing(WorkflowItemMessageRow::getText)
```
Comment (alexander-schmidt) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRow.java:97]: This change implies the assumption that there will always be at most one `WorkflowItemMessageRow` for any given `AGTransaction`. I wonder if that assumption is always fulfilled.

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java [hunk 5]
```diff
@@ -26,16 +29,41 @@ public WorkflowItemMessageRowsLoader(AGWorkflowItem item) {
 
 	public List<WorkflowItemMessageRow> loadAllRows() {
```
Comment (alexander-schmidt) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java:30]: I think we need `WorkflowItemMessageRowsLoaderTest`.

**Review by alexander-schmidt — DISMISSED** [review]

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java (hunk 1 above)
Comment (oliver-richers) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java:46]: done

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java (hunk 2 above)
Comment (oliver-richers) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java:41]: done

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRow.java (hunk 4 above)
Comment (oliver-richers) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRow.java:97]: No, that is not the point, there can be more than one `WorkflowItemMessageRow` per `AGTransaction`. This is why the loading order is important. `WorkflowItemMessageRow` records with the same `AGTransaction` need to be order like this:
- comments and tasks
- auto transition source states
- item logs

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java (hunk 5 above)
Comment (oliver-richers) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/message/WorkflowItemMessageRowsLoader.java:30]: that will take 1h or 2h, so, will not do today

**Review by daniel-klose — APPROVED** [review]

**Pull request merged**